plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/pvc/CommonPVCStrategyWithEphemeralWorkspacesSupport.java

```diff
@@ -0,0 +1,164 @@
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.workspace.config.Volume;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.WorkspaceManager;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceImpl;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.Names;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespaceFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.CommonPVCStrategy;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.PVCSubPathHelper;
+
+/**
+ * Extends {@code CommonPVCStrategy} and provides support of the ephemeral workspaces based on the
+ * `emptyDir` volumes. If workspace config contains `mountSources` attribute which is set to true,
+ * than instead of using regular {@code CommonPVCStrategy}, each volume would be represented as an
+ * `emptyDir`. NOTE: no data would be persisted once between ephemeral workspace restarts, since
+ *
+ * @see <a href="https://kubernetes.io/docs/concepts/storage/volumes/#emptydir">EmptyDir</a>
+ * @author Ilya Buziuk
+ */
+public class CommonPVCStrategyWithEphemeralWorkspacesSupport extends CommonPVCStrategy {
+  private static final String MOUNT_SOURCES_ATTRIBUTE = "mountSources";
+
+  private final WorkspaceManager workspaceManager;
+  private final String volumeNamePrefix;
+
+  @Inject
+  public CommonPVCStrategyWithEphemeralWorkspacesSupport(
+      @Named("che.infra.kubernetes.pvc.name") String pvcName,
+      @Named("che.infra.kubernetes.pvc.quantity") String pvcQuantity,
+      @Named("che.infra.kubernetes.pvc.access_mode") String pvcAccessMode,
+      @Named("che.infra.kubernetes.pvc.precreate_subpaths") boolean preCreateDirs,
+      PVCSubPathHelper pvcSubPathHelper,
+      KubernetesNamespaceFactory factory,
+      WorkspaceManager workspaceManager) {
+    super(pvcName, pvcQuantity, pvcAccessMode, preCreateDirs, pvcSubPathHelper, factory);
+    this.workspaceManager = workspaceManager;
+    this.volumeNamePrefix = pvcName;
+  }
+
+  @Override
+  public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+    String workspaceId = identity.getWorkspaceId();
+    if (isEphemeral(workspaceId)) {
+      for (Pod pod : k8sEnv.getPods().values()) {
+        final PodSpec podSpec = pod.getSpec();
+        for (Container container : podSpec.getContainers()) {
+          final String machineName = Names.machineName(pod, container);
+          Map<String, Volume> volumes = k8sEnv.getMachines().get(machineName).getVolumes();
+          addMachineVolumes(workspaceId, pod, container, volumes);
```

Comment:
What will happen with volumes that might come from the WS recipe or provisioned by the infrastructure?

Reply:
Our infra can add some additional stuff and PVC strategies should respect it. We have just one such case AFAIK at the moment - logs volume. But it is workarounded in existing strategies (including the strategy from this PR). 
Maybe it is OK and infra should not add anything in addition. Not sure. For the time being this code should work I suppose. 
@sleshchenko might want to comment after he is back from a sick leave. 
@ibuziuk consider this thread as thoughts out loud, it doesn't block the PR